daffodil-runtime2/src/main/c/common_runtime.c

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```

Comment:
A thought, instead of having this cc plugin and putting code in src/main/c/ to compile everything into a static lib, what if we instead just put these c/h files in src/main/resources/, and don't actually compile them as part of the sbt build. Instead when we generate code for a DFDL schema, we just copy these files out of the jar and then compile them along with the generate files? This removes any complexity in the sbt config about cc or libraries or anything, and just treats the static stuff like normal resources, making the sbt config much simpler.

Downside is it would perhaps take a little longer to compile generated code since every schema will need to compile these same files, but perhaps the compilation is fast enough? And I suspect as this is made to support more complex schemas, the size of the generated code will outnumber the size of the static stuff. And if compilation time was an issue, the compiling part of this code generator could separately build the static lib and cache it for reuse for other DFDL schemas.

Another downisde is that you can't easily compile and test just the static parts, but I figure the bulk of testing needs to be done with the generated code to ensure correctness, which will implicitly test the static code.

Reply:
Eventually, this "static code" is going to become a rather large library. As big as both runtime1 parser and unparser together. Bigger given that this is C code, in terms of Lines of code of course. Think in terms of 50K lines of code kind of library. So I think it needs to be built as a static library. 

I am somewhat surprised that just putting c code into src/main/c  and /src/main/h doesn't just work transparently. We can't be the first project mixing C code and Java/scala using SBT. Lots of JNI project must do this right?

Reply:
Are you suggesting this would be too large to compile along with the generated code?

Looks like JNI code includes compiled libraries in the jar and then uses System.load to load the libraries:

https://stackoverflow.com/questions/2937406/how-to-bundle-a-native-library-and-a-jni-library-inside-a-jar

We can probably distribute the static library in the jar, and then the c-generator can extract it when needed. I don't think there's  a need for System.load since the runtime isn't actually executing that code. But that suggests the right distribution medium is withing the jar. The C headers still need to be included in the the jar  and extracted to allow the generated code to be compiled with it though.

Reply:
I'd rather not get in the bussiness of shipping precompiled static libraries.

At the moment, Daffodil is mostly platform independent by virtue of running on the JVM. This goes away if we include compiled c code. Further, providing a shared library for multiple supported systems from a single build would involve setting up cross compilation, which is another headache.

Since our output is c source code anyway, the user already needs to go through a c compile step regardless, so I do not think we are saving them much by pre compiling the common runtime.

Further, a common usecase for runtime2 would likely be embedded systems, where space is at a premium. Users of such systems may prefer to compile everything at once with static linking in order to benifit from the size reductions of dead code elimination (as portions of the runtime may not be used by every format)

If we do want to provide a shared library, I don't think including it in the jar makes sense. You need the .so at parser runtime; which (unlike with jni) does not occur under a jvm. If we provide a .so, I would think it should be distributed seperatly.

Reply:
By "shared" library, I did not intend to suggest ".so" vs. ".a", but just "shared" generally, as in not need to be duplicated. I do understand many users would want ".a" type linkage. 

I think we want people to be able to run TDML tests against the C-generating backend. Leveraging TDML tests for the overlapping portable subset of DFDL that both backends support is critical to success. 

That suggests we need some "on the fly" way of compiling, linking, and running the C code. 

As for packaging for "real" usage like when the user just wants to complile their DFDL schema and take away the C-code to some other system, I'm less concerned about this. Just creating a zip/jar of the include and C-code directories would seem sufficient to me. But we do want it to be 100% clear what part is generated, and so if they recompile the schema, would be discarded and overwritten by new generated code. That needs to be distinguished from the static "library" code (and its headers) which would not be regenerated.